src/binder/bind/bind_reading_clause.cpp

```diff
@@ -137,86 +137,136 @@ std::unique_ptr<BoundReadingClause> Binder::bindLoadFrom(
+    // Bind columns from input.
+    std::vector<std::string> expectedColumnNames;
+    std::vector<std::unique_ptr<common::LogicalType>> expectedColumnTypes;
     for (auto& [name, type] : loadFrom.getColumnNameDataTypesRef()) {
-        inputColumnNames.push_back(name);
-        inputColumnTypes.push_back(bindDataType(type));
+        expectedColumnNames.push_back(name);
+        expectedColumnTypes.push_back(bindDataType(type));
     }
+    // Detect columns from file.
     std::vector<std::string> detectedColumnNames;
     std::vector<std::unique_ptr<common::LogicalType>> detectedColumnTypes;
-    switch (fileType) {
+    sniffFiles(*readerConfig, detectedColumnNames, detectedColumnTypes);
+    // Validate and resolve columns to use.
+    expression_vector columns;
+    if (expectedColumnTypes.empty()) { // Input is empty. Use detected columns.
+        columns = createColumnExpressions(*readerConfig, detectedColumnNames, detectedColumnTypes);
+    } else {
+        validateNumColumns(expectedColumnTypes.size(), detectedColumnTypes.size());
+        if (fileType == common::FileType::PARQUET) {
+            validateColumnTypes(expectedColumnNames, expectedColumnTypes, detectedColumnTypes);
+        }
+        columns = createColumnExpressions(*readerConfig, expectedColumnNames, expectedColumnTypes);
+    }
+    auto info = std::make_unique<BoundFileScanInfo>(
+        std::move(readerConfig), std::move(columns), nullptr, TableType::UNKNOWN);
+    auto boundLoadFrom = std::make_unique<BoundLoadFrom>(std::move(info));
+    if (loadFrom.hasWherePredicate()) {
+        auto wherePredicate = expressionBinder.bindExpression(*loadFrom.getWherePredicate());
+        boundLoadFrom->setWherePredicate(std::move(wherePredicate));
+    }
+    return boundLoadFrom;
+}
+
+expression_vector Binder::createColumnExpressions(common::ReaderConfig& readerConfig,
+    const std::vector<std::string>& columnNames,
+    const std::vector<std::unique_ptr<common::LogicalType>>& columnTypes) {
+    expression_vector columns;
+    for (auto i = 0u; i < columnTypes.size(); ++i) {
+        auto columnName = columnNames[i];
+        auto columnType = columnTypes[i].get();
+        readerConfig.columnNames.push_back(columnName);
+        readerConfig.columnTypes.push_back(columnType->copy());
+        columns.push_back(createVariable(columnName, *columnType));
+    }
+    return columns;
+}
+
+void Binder::validateColumnTypes(const std::vector<std::string>& columnNames,
+    const std::vector<std::unique_ptr<LogicalType>>& expectedColumnTypes,
+    const std::vector<std::unique_ptr<LogicalType>>& detectedColumnTypes) {
+    assert(expectedColumnTypes.size() == detectedColumnTypes.size());
+    for (auto i = 0; i < expectedColumnTypes.size(); ++i) {
+        if (*expectedColumnTypes[i] != *detectedColumnTypes[i]) {
+            throw BinderException(
+                StringUtils::string_format("Column {} type mismatch. Expect {} but get {}.",
```

Comment:
```suggestion
                StringUtils::string_format("Column {} type mismatch. Expected {} but got {}.",
```

Our codebase has this issue a lot. The correct English error message is always:

```
Expected {} but got {}.
```

Always past tense.

Reply:
Also, let's do:

```
Column `{}` type mismatch.
```

Or something. The error message in the test (Column name type mismatch) is hard to read :P